src/wallet/wallet2.cpp

```diff
@@ -1183,7 +1184,7 @@ wallet2::wallet2(network_type nettype, uint64_t kdf_rounds, bool unattended):
   m_light_wallet_balance(0),
   m_light_wallet_unlocked_balance(0),
   m_original_keys_available(false),
-  m_message_store(),
+  m_message_store(std::unique_ptr<epee::net_utils::http::abstract_http_client>(m_http_client->clone())),
```

Comment:
Consider to use `std::shared_ptr` instead

Reply:
A shared pointer would introduce a functional change from current master, which uses a separate http client instance for each class.  The functional tests fail when changed to a shared pointer.

Reply:
I don't like `virtual abstract_http_client* clone() const = 0;` idea.

Reply:
The message store needs to be initialized with a http client of the same type as wallet2, which the clone() achieves.  Do you suggest any different or better approach?

Reply:
By the way, a shared_ptr was my first choice as well.

Reply:
I don't understand why switching to a shared_ptr instead of unique_ptr would break (I don't care which one is used otherwise). Can you expand on this a bit ?

Reply:
Clone implies we are copying some existing instance state. But actually the message store just needs a new http client instead (not a clone of an existing one).

Off the top of my head, you could pass http client factory to wallet2 and message store.

In the simplest case it could be just a callback returning new `std::unique_ptr<abstract_http_client>` instance.

Reply:
Ah, i didn't see your guys comments before posting my reply.

Reply:
Ok I added a factory and removed the clone().  I can make name and styling change if there are any suggestions.